### PR TITLE
Allow setting the domainURIPrefix for custom domain names/paths when creating dynamic links

### PR DESCRIPTION
**Pull request by dmandar**
Hey there! So you want to contribute to a Firebase SDK?
Before you file this pull request, please read these guidelines:

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here.
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help
    us make Firebase APIs better, please propose your change in a feature request so that we
    can discuss it together.

### CODE REVIEW

File: Firebase/DynamicLinks/Public/FDLURLComponents.h [hunk 1]
```diff
@@ -517,6 +517,9 @@ FIR_SWIFT_NAME(DynamicLinkComponents)
 + (instancetype)componentsWithLink:(NSURL *)link
                             domain:(NSString *)domain
     NS_SWIFT_UNAVAILABLE("Use init(link:domain:)");
+DEPRECATED_MSG_ATTRIBUTE(
```
Comment (paulb777) [inline Firebase/DynamicLinks/Public/FDLURLComponents.h]: Fix the build warnings caught - https://travis-ci.org/firebase/firebase-ios-sdk/jobs/454524370 - See other examples for how to use DEPRECATE_MSG_ATTRIBUTE.

**Comment by paulb777** [comment]
- WARN  | xcodebuild:  /Users/travis/build/firebase/firebase-ios-sdk/Firebase/DynamicLinks/Public/FDLURLComponents.h:520:1: warning: declaration does not declare anything [-Wmissing-declarations]
    - WARN  | xcodebuild:  /Users/travis/build/firebase/firebase-ios-sdk/Firebase/DynamicLinks/Public/FDLURLComponents.h:534:1: warning: declaration does not declare anything [-Wmissing-declarations]
    - WARN  | xcodebuild:  /Users/travis/Library/Developer/Xcode/DerivedData/App-anfynblkxhfadcdawyblosijbtmb/Build/Products/Release-iphonesimulator/FirebaseDynamicLinks/FirebaseDynamicLinks.framework/Headers/FDLURLComponents.h:520:1: warning: declaration does not declare anything [-Wmissing-declarations]
    - WARN  | xcodebuild:  /Users/travis/Library/Developer/Xcode/DerivedData/App-anfynblkxhfadcdawyblosijbtmb/Build/Products/Release-iphonesimulator/FirebaseDynamicLinks/FirebaseDynamicLinks.framework/Headers/FDLURLComponents.h:534:1: warning: declaration does not declare anything [-Wmissing-declarations]

File: Example/DynamicLinks/FDLBuilderTestAppObjC/ViewController.m [hunk 2]
```diff
@@ -289,7 +289,7 @@ - (void)_initDefaultValues {
 - (void)_buildFDLLink {
   NSURL *link = [NSURL URLWithString:_paramValues[@"linkString"]];
   FIRDynamicLinkComponents *components =
-      [FIRDynamicLinkComponents componentsWithLink:link domain:_paramValues[@"domain"]];
+      [FIRDynamicLinkComponents componentsWithLink:link domainURIPrefix:_paramValues[@"domain"]];
```
Comment (mh6o) [inline Example/DynamicLinks/FDLBuilderTestAppObjC/ViewController.m]: _paramValues[@"domainUriPrefix"] ?

File: Example/DynamicLinks/Tests/FDLURLComponentsTests.m [hunk 3]
```diff
@@ -499,7 +516,7 @@ - (void)testFDLComponentsFailsOnMalformedDomain {
   NSURL *link = [NSURL URLWithString:linkString];
 
   FIRDynamicLinkComponents *components =
-      [FIRDynamicLinkComponents componentsWithLink:link domain:@"this is invalid domain"];
+      [FIRDynamicLinkComponents componentsWithLink:link domainURIPrefix:@"this is invalid domain"];
```
Comment (mh6o) [inline Example/DynamicLinks/Tests/FDLURLComponentsTests.m]: nit, this is invalid domain URL prefix

File: Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m [hunk 4]
```diff
@@ -448,15 +449,37 @@ - (instancetype)init {
...
 }
 
 - (instancetype)initWithLink:(NSURL *)link domain:(NSString *)domain {
+  NSString *domainURIPrefix = [NSString stringWithFormat:@"https://%@", domain];
+  return [self initWithLink:link domainURIPrefix:domainURIPrefix];
+}
+
+#pragma mark Initializers.
++ (instancetype)componentsWithLink:(NSURL *)link domainURIPrefix:(NSString *)domainURIPrefix {
+  if (![[domainURIPrefix lowercaseString] hasPrefix:@"https://"]) {
```
Comment (mh6o) [inline Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: Can we perform stricter checks up-front, not just prefix? There's a well-defined regex that determines what qualifies as a URL.

File: Example/DynamicLinks/FDLBuilderTestAppObjC/ViewController.m (hunk 2 above)
Comment (dmandar) [reply Example/DynamicLinks/FDLBuilderTestAppObjC/ViewController.m]: Done.

File: Example/DynamicLinks/Tests/FDLURLComponentsTests.m (hunk 3 above)
Comment (dmandar) [reply Example/DynamicLinks/Tests/FDLURLComponentsTests.m]: Done.

File: Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m (hunk 4 above)
Comment (dmandar) [reply Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: Done.

File: Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m [hunk 5]
```diff
@@ -448,15 +449,50 @@ - (instancetype)init {
 
 @implementation FIRDynamicLinkComponents
 
+#pragma mark Deprecated Initializers.
 + (instancetype)componentsWithLink:(NSURL *)link domain:(NSString *)domain {
-  return [[self alloc] initWithLink:link domain:domain];
+  NSString *domainURIPrefix = [NSString stringWithFormat:@"https://%@", domain];
```
Comment (ryanwilson) [inline Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: Is there any chance that a developer could already pass in a domain with `https://` in front? What happens if the site doesn't support `https`?

File: Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m [hunk 6]
```diff
@@ -448,15 +449,50 @@ - (instancetype)init {
...
+  return [FIRDynamicLinkComponents componentsWithLink:link domainURIPrefix:domainURIPrefix];
 }
 
 - (instancetype)initWithLink:(NSURL *)link domain:(NSString *)domain {
+  NSString *domainURIPrefix = [NSString stringWithFormat:@"https://%@", domain];
```
Comment (ryanwilson) [inline Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: Same as above.

File: Firebase/DynamicLinks/Public/FDLURLComponents.h [hunk 7]
```diff
@@ -516,7 +516,9 @@ FIR_SWIFT_NAME(DynamicLinkComponents)
  */
 + (instancetype)componentsWithLink:(NSURL *)link
                             domain:(NSString *)domain
-    NS_SWIFT_UNAVAILABLE("Use init(link:domain:)");
+    NS_SWIFT_UNAVAILABLE("Use init(link:domain:)")DEPRECATED_MSG_ATTRIBUTE(
+        "This method is deprecated. Please use the new method with support for "
+        "domainUriPrefix- componentsWithLink:domainURIPrefix.");
```
Comment (ryanwilson) [inline Firebase/DynamicLinks/Public/FDLURLComponents.h]: `domainURIPrefix`

File: Firebase/DynamicLinks/Public/FDLURLComponents.h [hunk 8]
```diff
@@ -527,7 +529,39 @@ FIR_SWIFT_NAME(DynamicLinkComponents)
  * @param domain Domain of your App. This value must be equal to your assigned domain from Firebase
  *     Console.
  */
-- (instancetype)initWithLink:(NSURL *)link domain:(NSString *)domain;
+- (instancetype)initWithLink:(NSURL *)link
+                      domain:(NSString *)domain
+    DEPRECATED_MSG_ATTRIBUTE(
+        "This method is deprecated. Please use the new method with support for "
+        "domainUriPrefix- initWithLink:domainURIPrefix.");
```
Comment (ryanwilson) [inline Firebase/DynamicLinks/Public/FDLURLComponents.h]: URI instead of Uri

File: Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m (hunk 5 above)
Comment (dmandar) [reply Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: We mandate https as the only acceptable scheme (as decided in api review doc). The previous methods always had the domain name only  (without scheme), but valid point that developers might sometimes get confused now. Added extra check.

Comment (dmandar) [reply Firebase/DynamicLinks/FDLURLComponents/FDLURLComponents.m]: Done.

**Pull request merged**